File: No.15368/Answer3.cs
using System;

class Answer{
    static void Main(string[] args)
    {
        if(int.TryParse(Console.ReadLine(), out int n)){
            int[] par = Array.ConvertAll(("0 "+Console.ReadLine()).Split(" "), s => int.Parse(s));
            ulong[] answerValues = new ulong[n + 1];
            int[] employeeChildValue = new int[n + 1];

            for(int i = n; i <= 1; i--){
                answerValues[par[i]] += ++answerValues[i] + (ulong)++employeeChildValue[i];
                employeeChildValue[par[i]] += employeeChildValue[i];
            }

            for(int i = 0; i < answerValues.Length; i++){
                Console.Write(answerValues[i] + " ");
            }
        }
    }
}

File: No.15368/Answer3.kt
import java.util.*

fun main(args: Array<String>) = with(Scanner(System.`in`)) {
   var n = nextLine().toInt();
   var par = ("0 0 "+  nextLine()).split(" ").map { it.toInt() }.toMutableList()
   var answerValues = MutableList<Long>(n + 1, {0})
   var employeeChildValue = MutableList<Int>(n + 1, {0})

   for (i in n downTo 1) {
      answerValues[par[i]] += ++answerValues[i] + ++employeeChildValue[i];
      employeeChildValue[par[i]] += employeeChildValue[i];
  }

   for(i in 1..n){
      print(answerValues[i].toString() + " ")
   }
}
